### ISSUE TITLE
Custom authorizer not being invoked

### ISSUE THREAD

**Issue by lewisbennett** [aws/aws-iot-device-sdk-java-v2#154]
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-iot-device-sdk-java-v2/issues) and didn't find any solution

**Describe the bug**
Lambda function connected to custom AWS IoT authorizer not being invoked by connection from Java device SDK.

**SDK version number**
V2

**Platform/OS/Hardware/Device**
Java, Windows 10

**To Reproduce (observed behavior)**
1) Create a custom AWS IoT authorizer, with basic required permissions, optional token signing **disabled**, and connect a Lambda function.

2) Add the following to the `builder` configuration part of the Shadow sample code (starts at line 236 in Shadow sample project):
```
// You also get the same result if you prefix the authorizer name with $. I'm not 100% sure which is the correct format, but the issue is present in either case.
.withUsername("testing?x-amz-customauthorizer-name=[name of custom authorizer]")
.withPassword("Password123")
```

3) Run the modified Shadow sample project with required parameters using port 443.

**Expected behavior**
I expect to see an invocation in the CloudWatch metrics section for the Lambda function to indicate that the authorizer and Lambda function were called correctly as a result of the incoming connection from the device SDK.

I am able to see invocation data when I test the authorizer using the AWS CLI, but never from the device SDK.

**Comment by lewisbennett** [comment]
I have managed to find a fix for the problem. This is not a long term solution, however.

If I modify the `AwsIotMqttConnectionBuilder` using Java's reflection API, I am able to invoke the custom authorizer. Please see the below snippet for reference.

```
try {
    Field field = builder.getClass().getDeclaredField("tlsOptions");

    field.setAccessible(true);

    TlsContextOptions options = (TlsContextOptions) field.get(builder);

    options.alpnList.clear();
    options.alpnList.add("mqtt");

} catch (NoSuchFieldException | IllegalAccessException e) {
    e.printStackTrace();
}
```

This snippet should be inserted between creating and configuring the connection builder, and building it into the final `MqttClientConnection` object.

The reason this works is because, by default, `TlsContextOptions.alpnList` contains `x-amzn-mqtt-ca`, which appears to configure the connection to communicate with AWS IoT core directly and, as a result, ignore custom authorizers.

Guidance on how to achieve the same result as above, or adding support to the SDK, would be appreciated.

Thanks.

**Comment by bretambrose** [comment]
Excellent detective work.  Looking over the builder there are a few simple ways we could address this:

1. add another constructor that takes TlsContextOptions and accepts them as gospel
2. add a mutator for the alpn list

Neither fix would help the next person who runs into this realize what the problem is (i.e. the work you did) though.

I'm wondering if it would be better to have a higher-level API on the builder, ala

withCustomAuthorizationConfig(...username..., ...password..., TlsOptions)

or similar that people in your situation would naturally gravitate to and hopefully avoid running into this.

Do you have any suggestions here for what you would find intuitive/useful?

**Comment by lewisbennett** [comment]
Thanks for your reply.

While looking into this I spent a fair amount of time staring at the names of the mutator methods to see if I was just overlooking something, so I think this would be the way to go.

To avoid confusion with having multiple methods that reference the username and password, I would keep them separate and have a method similar to the following:

```
public AwsIotMqttConnectionBuilder withCustomAuthorizerConfig(String authorizerName, String signature, String tokenName)
```

The idea here is that adding the query parameters to the username is handled by the SDK for you when the user calls `AwsIotMqttConnectionBuilder.build()`, allowing the user to only have to worry about providing the values. In the case of `signature` and `tokenName`, these would be optional. This can also tie into the `resetDefaultPort` method, since custom authorizers are only supported over port 443. It would also allow the SDK to handle configuring the ALPN list correctly too, although there may be advantages to making this ability public as well.

The only problem I see with the above is that the [documentation](https://docs.aws.amazon.com/iot/latest/developerguide/custom-auth.html) does state that custom authorizer configuration is specified using query parameters on the MQTT username. This may entice users to do what I've done and add them to the `.withUsername(...)` method manually. That said, the documentation isn't giving an example for the Java device SDK, it's more of a general approach, so may not be a problem.

Open to suggestions and feedback on the idea. Also, I haven't had to use websockets for my implementation, so I don't know if the above would affect them in any way.

**Comment by bretambrose** [comment]
If you're interested, feel free to create a PR with what you envision.  I will try and get to this probably late next week, but oncall may interfere.

**Comment by lewisbennett** [comment]
Plot thickens.

After doing some more testing, it would seem that the SDK doesn't actually call the custom authorizer. At the time of posting my workaround, I had the authorizer set as the account default. This was done in an attempt to troubleshoot the issues I was facing previously.

Thinking that I had found the problem and come up with a temporary workaround, I cleared the account default custom authorizer and the original problem returned (i.e. the custom authorizer isn't invoked when specified with the query parameter).

I will try and debug this further when I have the time.

**Comment by TheNodi** [comment]
@lewisbennett I don't know if you have any update on this, but I think I run into your same problem.

I'm using a custom authorizer, with request signing, without default authorizer configured.
After applying the workaround in https://github.com/aws/aws-iot-device-sdk-java-v2/issues/154#issuecomment-815862213 , the authorizer is still not invoked.

Since the username is a query string, the SDK is appending debug / tracking information to it:
https://github.com/aws/aws-iot-device-sdk-java-v2/blob/6b711b13d80b62baf7772ff1a1739b1fae346170/sdk/src/main/java/software/amazon/awssdk/iot/AwsIotMqttConnectionBuilder.java#L525

In the builder, there's no check for existing query strings. The resulting (invalid) username is:
```
username?x-amz-customauthorizer-name=my-authorizer&x-amz-customauthorizer-signature=....?SDK=JavaV2&Version=0.15
```
(Note the double `?` after `username` and before `SDK`)

IoT Core is rejecting the request, probably because it is failing the parsing of the username. If I use the debugger to replace the second `?` with `&`, the custom authorizer is invoked correctly.
I think the simple and immediate solution to this is check whether the username is a query string, and choose the appropriate concatenation character.

**Comment by lewisbennett** [comment]
Hi, unfortunately there's nothing new from my side. The project we were working on was put on hold shortly after we were investigating this problem. I do however, remember seeing the double `?` somewhere.

I agree with your solution, more so because of the fact that there aren't any methods for custom authorizers to define behaviour in this way (if I recall correctly).

**Comment by TwistedTwigleg** [comment]
This should be fixed in version `v1.8.5`. Please give it a try (there's also a new sample) and if the problem persists, please make another issue. Thanks!
Closing this issue...
